tiledb/api/src/array/enumeration.rs

```diff
@@ -0,0 +1,502 @@
+                self.context.capi(),
+                self.capi(),
+                &mut ptr as *mut *const std::ffi::c_uchar
+                    as *mut *const std::ffi::c_void,
+                &mut size,
+            )
+        };
+
+        // Rust docs say that we're not allowed to pass a nullptr to the
+        // std::slice::from_raw_parts because that breaks how Option<&[T]>::None
+        // is represented.
+        if ptr.is_null() {
+            ptr = 0x1 as *const std::ffi::c_uchar;
+        }
+
+        if res == ffi::TILEDB_OK {
+            let slice =
+                unsafe { std::slice::from_raw_parts(ptr, size as usize) };
+            Ok(slice)
+        } else {
+            Err(self.context.expect_last_error())
+        }
+    }
+
+    pub fn offsets(&self) -> TileDBResult<&[u64]> {
+        let mut ptr: *const std::ffi::c_ulonglong = out_ptr!();
+        let mut size: u64 = 0;
+        let res = unsafe {
+            ffi::tiledb_enumeration_get_offsets(
+                self.context.capi(),
+                self.capi(),
+                &mut ptr as *mut *const std::ffi::c_ulonglong
+                    as *mut *const std::ffi::c_void,
+                &mut size,
+            )
+        };
+
+        // Rust docs say that we're not allowed to pass a nullptr to the
+        // std::slice::from_raw_parts because that breaks how Option<&[T]>::None
+        // is represented.
+        if ptr.is_null() {
+            ptr = 0x1 as *const std::ffi::c_ulonglong;
+        }
+
+        // The size returned is the number of bytes, where from_raw_parts
+        // wants the number of elements.
+        assert!(size as usize % std::mem::size_of::<u64>() == 0);
+        let elems = size as usize / std::mem::size_of::<u64>();
+
+        if res == ffi::TILEDB_OK {
+            let slice = unsafe { std::slice::from_raw_parts(ptr, elems) };
+            Ok(slice)
+        } else {
+            Err(self.context.expect_last_error())
+        }
+    }
+
+    pub fn extend(&self, column: Column) -> TileDBResult<Enumeration<'ctx>> {
+        let mut c_new_enmr: *mut ffi::tiledb_enumeration_t = out_ptr!();
+
+        // Rust semantics require that slice pointers aren't nullptr so that
+        // nullptr can be used to distinguish between Some and None. The stdlib
+        // empty slices all appear to return 0x1 which is mentioned in the docs
+        // as a valid strategy. For our situation, we just use a zero length to
+        // indicate when we should pass nullptr.
+        let offsets_ptr = if column.offsets().is_empty() {
+            std::ptr::null_mut()
+        } else {
+            column.offsets().as_ptr()
+        };
+
+        // An important note here is that the Enumeration allocator copies the
+        // contents of data of offsets rather than assumes ownership. That
+        // means this is safe as those bytes are guaranteed to be alive until
+        // we drop self at the end of this method after returning from
```

Comment:
Is this method meant to consume `self`?  RIght now it is a reference

Reply:
It is specifically not consuming self because of the underlying implementation. `tiledb::sm::Enumeration::extend` creates a new instance of `Enumeration` that gets a copy of the "parent" enumeration's data with a copy of the extension data appended. I.e., it's a copy underneath rather than a sharing of data.

The reasoning here is that this function was added to aid language bindings do schema evolution so the "parent" instance is still in use in its original form which is what prompts the returning a new instance design.

Reply:
Oh, that makes sense - I was confused because the line break starts with "we drop self at the end of this method", but you don't.

Does this mean you end up with two enumerations with the same name?  Are they also versioned like the array writes are?

Reply:
Correct, the background here is that once users got access to Enumerations, the immediately starting asking to allow for empty Enumerations that are then evolved to have more values. The general idea was that the actual enumeration values aren't known at Array creation time, and are instead only discovered during the first import.

The way it works is before a write, users can evolve the schema with the new Enumeration values. However, the way schema evolution works is basically that we just write a new one and then all new writes will use that instead of the older versions. So the original Enumeration stays for the fragments written using it, and then the new schema has a new copy with the same name (but on disk its a UUID mapping so they don't clash). There's also logic that ensures that names in a given schema are unique, but not necessarily across multiple schemas.

Also, for extra color, the entire extend API is also just a helper to avoid a foot gun in that the values of the prefix must be equivalent for the schema evolution to succeed. Without that you would end up where completely different values are returned depending on which fragments were opened for reading.